third_party/move/move-compiler-v2/src/pipeline/unreachable_code_analysis.rs

```diff
@@ -0,0 +1,157 @@
+//! This annotation is a prerequisite for the unreachable code checker and unreachable code remover.
+//!
+//! This analysis a forward "may" analysis, it tracks whether an instruction is:
+//! - maybe reachable (there may be an execution path from the function entry to the instruction)
+//! - definitely not reachable (there is no execution path from the function entry to the instruction)
+
+use move_binary_format::file_format::CodeOffset;
+use move_model::model::FunctionEnv;
+use move_stackless_bytecode::{
+    dataflow_analysis::{DataflowAnalysis, TransferFunctions},
+    dataflow_domains::{AbstractDomain, JoinResult},
+    function_target::{FunctionData, FunctionTarget},
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    stackless_bytecode::Bytecode,
+    stackless_control_flow_graph::StacklessControlFlowGraph,
+};
+use std::collections::BTreeMap;
+
+/// Reachability state of an instruction.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum ReachableState {
+    Maybe, // Maybe reachable from function entry
+    No,    // Definitely not reachable from function entry
+}
+
+impl ReachableState {
+    /// Mark this state as maybe reachable from the function entry.
+    fn mark_as_maybe_reachable(&mut self) {
+        *self = ReachableState::Maybe;
+    }
+
+    /// Mark this state as definitely not reachable from the function entry.
+    fn mark_as_not_reachable(&mut self) {
+        *self = ReachableState::No;
+    }
+}
+
+impl AbstractDomain for ReachableState {
+    fn join(&mut self, other: &Self) -> JoinResult {
+        use ReachableState::*;
+        match (self.clone(), other) {
+            (No, Maybe) => {
+                self.mark_as_maybe_reachable();
+                JoinResult::Changed
+            },
+            (Maybe, _) | (No, No) => JoinResult::Unchanged,
+        }
+    }
+}
+
+/// Mapping from code offsets to their reachability state, before executing the
+/// instruction at the code offset.
+#[derive(Clone, Debug)]
+pub struct ReachableStateAnnotation(BTreeMap<CodeOffset, ReachableState>);
+
+impl ReachableStateAnnotation {
+    /// Is the instruction at the given `offset` definitely not reachable?
+    pub fn is_not_reachable(&self, offset: CodeOffset) -> bool {
+        self.0
+            .get(&offset)
+            .map_or(true, |state| matches!(state, ReachableState::No))
+    }
+}
+
+/// Forward intra-procedural dataflow analysis.
+/// Determines whether an instruction is reachable or not.
+pub struct UnreachableCodeAnalysis {}
+
+impl UnreachableCodeAnalysis {
+    /// Analyze the given function and return a mapping from code offsets to their reachability state.
+    fn analyze(&self, func_target: &FunctionTarget) -> ReachableStateAnnotation {
+        let code = func_target.get_bytecode();
+        let cfg = StacklessControlFlowGraph::new_forward(code);
+        // We assume the entry of a function is reachable, as we have implemented this analysis
```

Comment:
This comment maybe more confusing than helpful. Indeed this is kind of analysis is local to a function. I don't think it is an implementation choice, or is it?

Also not sure to what it relates, below you start from Maybe (false) as the bottom element, with your join being logical OR.

Reply:
> I don't think it is an implementation choice, or is it?

We could further enhance the precision of this analysis by extending it to be inter-procedural. For example, in the test case `inter_procedural_abort.move` that I have added:

```
module 0xc0ffee::m {
    fun always_abort() {
        abort 0
    }

    fun test(): u64 {
        always_abort();
        let i = 0; // interprocedural analysis can prove this to be unreachable, intra cannot.
        i = i + 1;
        i
    }
}
```

I will remove the mention of inter vs. intra in this comment though, as it seems to cause some confusion.

Reply:
I see what you meant. However, this is rather unusual and not part of standard unreachable code analysis in PLs.